### PR TITLE
Support 3 inputs models in rerank calculator

### PR DESCRIPTION
**Pull request by michalkulakowski**
### 🛠 Summary

CVS-171211

### 🧪 Checklist

- [ ] Unit tests added.
- [ ] The documentation updated.
- [ ] Change follows security best practices.
``

### CODE REVIEW

File: src/sidepacket_servable.cpp [hunk 1]
```diff
@@ -33,9 +33,17 @@
...
-        token = modelConfig[token_id_name].GetInt64();                                  \
+#define SET_TOKEN_ID(token, token_id_name)                                                                                            \
+    if (modelConfig.HasMember(token_id_name) && modelConfig[token_id_name].IsInt64() && modelConfig[token_id_name].GetInt64() != 0) { \
+        token = modelConfig[token_id_name].GetInt64();                                                                                \
+    }
+
+#define SET_TOKEN(token)                                                                   \
+    if (!token.has_value()) {                                                              \
+        if (tokenizerConfig.HasMember(#token) && tokenizerConfig[#token].IsString()) {     \
+            auto tokenizedInputs = tokenizer->encode(tokenizerConfig[#token].GetString()); \
```
Comment (mzegla) [inline src/sidepacket_servable.cpp]: why hashes?

File: src/rerank/rerank_calculator_ov.cc [hunk 2]
```diff
@@ -278,12 +284,17 @@ class RerankCalculatorOV : public CalculatorBase {
             // Prepare inputs for rerank model
             std::vector<size_t> chunk_mapping;
             auto [input_ids, attention_mask] = PrepareInputsForRerankModel(handler, chunk_mapping);
-
+            std::optional<ov::Tensor> typeIds;
+            if (rerank_session->getNumberOfModelInputs() == 3) {
+                typeIds = ov::Tensor{ov::element::i64, input_ids.get_shape()};
+                std::fill_n(typeIds->data<int64_t>(), input_ids.get_size(), 0);
```
Comment (mzegla) [inline src/rerank/rerank_calculator_ov.cc:290]: so we always fill that tensor with zeros?

File: src/sidepacket_servable.cpp [hunk 3]
```diff
@@ -81,6 +89,30 @@ SidepacketServable::SidepacketServable(const std::string& modelDir, const std::s
         SPDLOG_ERROR("Error during embeddings node plugin_config option parsing to JSON: {}", pluginConfig);
     }
     tokenizer = std::make_shared<ov::genai::Tokenizer>(parsedModelsPath);
+    std::filesystem::path tokenizerConfigPath = (std::filesystem::path(graphPath) / fsModelsPath / "tokenizer_config.json");
```
Comment (mzegla) [inline src/sidepacket_servable.cpp:95]: will it work on windows as well?

File: src/sidepacket_servable.cpp [hunk 4]
```diff
@@ -81,6 +89,30 @@ SidepacketServable::SidepacketServable(const std::string& modelDir, const std::s
...
+    if (std::filesystem::exists(tokenizerConfigPath)) {
+        std::ifstream ifs(tokenizerConfigPath.string());
+        if (ifs.is_open()) {
+            rapidjson::Document tokenizerConfig;
+            rapidjson::IStreamWrapper isw(ifs);
+            rapidjson::ParseResult parseResult = tokenizerConfig.ParseStream(isw);
+            if (parseResult.Code()) {
+                SPDLOG_ERROR("Parsing config.json failed: {}", rapidjson::GetParseError_En(parseResult.Code()));
```
Comment (mzegla) [inline src/sidepacket_servable.cpp]: config.json -> tokenizer_config.json?

File: src/sidepacket_servable.cpp [hunk 5]
```diff
@@ -81,6 +89,30 @@ SidepacketServable::SidepacketServable(const std::string& modelDir, const std::s
...
+            } else {
+                SET_TOKEN(pad_token);
+                SET_TOKEN(eos_token);
+                SET_TOKEN(bos_token);
+                if (!sep_token.has_value()) {
+                    if (tokenizerConfig.HasMember("sep_token") && tokenizerConfig["sep_token"].IsString()) {
+                        auto tokenizedInputs = tokenizer->encode(tokenizerConfig["sep_token"].GetString());
```
Comment (mzegla) [inline src/sidepacket_servable.cpp:110]: don't you need to specify `skip_special_tokens=False` property? is it a valid, non-empty string?

File: src/sidepacket_servable.cpp [hunk 6]
```diff
@@ -81,6 +89,30 @@ SidepacketServable::SidepacketServable(const std::string& modelDir, const std::s
...
+                        sep_token = reinterpret_cast<int64_t*>(tokenizedInputs.input_ids.data())[0];
```
Comment (mzegla) [inline src/sidepacket_servable.cpp]: is it possible that we don't have any data here and try to access non existing element?

File: src/sidepacket_servable.cpp [hunk 7]
```diff
@@ -80,7 +91,35 @@ SidepacketServable::SidepacketServable(const std::string& modelDir, const std::s
     if (!status.ok()) {
         SPDLOG_ERROR("Error during embeddings node plugin_config option parsing to JSON: {}", pluginConfig);
     }
-    tokenizer = std::make_shared<ov::genai::Tokenizer>(parsedModelsPath);
+    ov::AnyMap tokenizerProperties = {{"add_special_tokens", false}};
```
Comment (dkalinowski) [inline src/sidepacket_servable.cpp:93]: why do we stop adding special tokens? did you check accuracy against other servings after that change?

File: src/sidepacket_servable.cpp [hunk 8]
```diff
@@ -80,7 +91,35 @@ SidepacketServable::SidepacketServable(const std::string& modelDir, const std::s
...
+    if (std::filesystem::exists(tokenizerConfigPath)) {
+        std::ifstream ifs(tokenizerConfigPath.string());
+        if (ifs.is_open()) {
+            rapidjson::Document tokenizerConfig;
+            rapidjson::IStreamWrapper isw(ifs);
+            rapidjson::ParseResult parseResult = tokenizerConfig.ParseStream(isw);
+            if (parseResult.Code()) {
+                SPDLOG_ERROR("Parsing tokenizer_config.json failed: {}", rapidjson::GetParseError_En(parseResult.Code()));
+            } else {
+                SET_TOKEN(pad_token);
```
Comment (dkalinowski) [inline src/sidepacket_servable.cpp:105]: isnt there access to such values from ov::genai::Tokenizer object itself? it already reads that config file during tokenizer construction

File: src/sidepacket_servable.cpp [hunk 9]
```diff
@@ -80,7 +91,35 @@ SidepacketServable::SidepacketServable(const std::string& modelDir, const std::s
...
+                SET_TOKEN(eos_token);
+                SET_TOKEN(bos_token);
+                if (!sep_token.has_value()) {
+                    if (tokenizerConfig.HasMember("sep_token") && tokenizerConfig["sep_token"].IsString()) {
+                        auto tokenizedInputs = tokenizer->encode(tokenizerConfig["sep_token"].GetString());
+                        if (tokenizedInputs.input_ids.get_size() == 1)
+                            sep_token = reinterpret_cast<int64_t*>(tokenizedInputs.input_ids.data())[0];
```
Comment (dkalinowski) [inline src/sidepacket_servable.cpp:112]: are we sure that input_ids output is always of type int64

**Review by mzegla — APPROVED** [review]

**Review by dkalinowski — APPROVED** [review]

**Comment by dtrawins** [comment]
@michalkulakowski  can you add new tested model to the documentation?
